File: pyBASCloudAPI/tests/setpoint_tests.py
import unittest
import time 
import datetime

from definitions import *

import pyBAScloudAPI as api

BCAPI = api.EntityContext(BASCLOUD_TEST_URL)


def errorHandler(e, j):
    raise e

def authenticate():
    currentDateTime = int(time.time())
    print("CURRENT DATETIME: ", currentDateTime)

    currentDateTimeString = datetime.datetime.fromtimestamp(currentDateTime).strftime("%FT%T.000Z")
    print("CURRENT DATETIME STR: ", currentDateTimeString)

    BCAPI.authenticateWithUserLogin(BASCLOUD_TEST_EMAIL, BASCLOUD_TEST_PASS)

    print("TOKEN: ", BCAPI.getToken())
    print("EXPIRATION DATE: ", datetime.datetime.fromtimestamp(BCAPI.getTokenExpirationDate()).strftime("%FT%T.000Z"))



class TestSetPointCollection(unittest.TestCase):

    def test_setpoint(self):
        print("\tRequesting all setpoints...")

        setpoints = BCAPI.getSetPointsCollection(BASCLOUD_TEST_TENANT_UUID, errorHandler=errorHandler)
        print("\t\tOK.")

        print("\tFound setpoints: ", len(setpoints[0]))

        self.assertTrue(len(setpoints[0]) >= 0) # may no set points available


class TestSingleSetPoint(unittest.TestCase):

    def test_setpoint(self):
        print("\tRequesting single setpoint with UUID...")

        setpoint = BCAPI.getSetPoint(BASCLOUD_TEST_TENANT_UUID, BASCLOUD_TEST_STEPOINT_UUID)
        print("\t\tOK.")

        print("\tSetpoint UUID: ", setpoint.uuid)
        print("\tSetpoint Value: ", setpoint.value)
        print("\tSetpoint timestamp: ", datetime.datetime.fromtimestamp(setpoint.timestamp).strftime("%FT%T.000Z"))
        print("\tSetpoint created at: ", datetime.datetime.fromtimestamp(setpoint.createdAt).strftime("%FT%T.000Z"))
        print("\tSetpoint updated at: ", datetime.datetime.fromtimestamp(setpoint.updatedAt).strftime("%FT%T.000Z"))

        self.assertEqual(setpoint.uuid, BASCLOUD_TEST_STEPOINT_UUID)
        self.assertEqual(setpoint.value, 2345.67)
        self.assertTrue(setpoint.timestamp > 0)
        self.assertTrue(setpoint.createdAt > 0)
        self.assertTrue(setpoint.updatedAt > 0)


class TestCreateAndDeleteSetPoint(unittest.TestCase):

    def test_setpoint(self):
        currentDateTime = int(time.time())

        new_set = BCAPI.createSetPoint(BASCLOUD_TEST_TENANT_UUID, BASCLOUD_TEST_DEVICE_UUID, 2345.67, currentDateTime)
        print("\t\tOK.")

        new_uuid = new_set.uuid
        print("\t\tUUID: ", new_uuid)

        self.assertEqual(new_set.value, 2345.67)
        self.assertEqual(new_set.timestamp, currentDateTime)
        # self.assertTrue(new_set.createdAt-(2*60*60) >= currentDateTime) // TODO ML: this is wrong at the moment, for some reason the backend returns timestamps -2hours
        # self.assertTrue(new_set.updatedAt-(2*60*60) >= currentDateTime)

        print("\tAgain Requesting all setpoints...") # Thus deactivating/inactivating them

        print("\tFrom: ", datetime.datetime.fromtimestamp(currentDateTime-(60*60*24)).strftime("%FT%T.000Z"))
        print("\tUntil: ", datetime.datetime.fromtimestamp(currentDateTime+(60*60*24)).strftime("%FT%T.000Z"))

        setpoints = BCAPI.getSetPointsCollection(BASCLOUD_TEST_TENANT_UUID, from_=currentDateTime-(60*60*24), until=currentDateTime+(60*60*24), errorHandler=errorHandler)
        print("\t\tOK.")

        print("\tFound setpoints: ", len(setpoints[0]))

        self.assertTrue(len(setpoints[0]) >= 1) # may no set points available


if __name__ == '__main__':
    authenticate()
    unittest.main()

File: tests/setpoint_tests.cpp
#include <chrono>
#include <iostream>

#include <fmt/core.h>
#include <fmt/chrono.h>

#include <gtest/gtest.h>

#include "EntityContext.h"
#include "entity/EntityCollection.h"
#include "error/Exceptions.h"


using namespace BAScloud;

EntityContext BCAPI(BASCLOUD_TEST_URL);


void authenticate() {

    std::time_t currentDateTime = std::chrono::duration_cast<std::chrono::seconds>(std::chrono::system_clock::now().time_since_epoch()).count();

    std::cerr <<"CURRENT DATETIME: "<<fmt::format("{:%Y-%m-%d %H:%M:%S}", fmt::localtime(currentDateTime))<<std::endl;

    BCAPI.authenticateWithUserLogin(BASCLOUD_TEST_EMAIL, BASCLOUD_TEST_PASS);

    std::cerr <<"TOKEN: "<<BCAPI.getToken()<<std::endl;
    std::cerr <<"EXPIRATION DATE: "<<fmt::format("{:%Y-%m-%d %H:%M:%S}", fmt::localtime(BCAPI.getTokenExpirationDate()))<<std::endl;
}

TEST(BasicTests, SetPointCollectionTest) {

    std::cout << "\tRequesting all setpoints..." << std::endl;

    EntityCollection<SetPoint> setpoints = BCAPI.getSetPointsCollection(BASCLOUD_TEST_TENANT_UUID, {}, -1, -1, -1, -1, {}, -1, -1, [](const std::string& e, const std::string& j) {
            throw e;
        });
    std::cout << "\t\tOK." << std::endl;

    std::cout << "\tFound setpoints: " << setpoints.first.size() << std::endl;

    EXPECT_TRUE(setpoints.first.size() >= 0); // may no set points available
}

TEST(BasicTests, SingleSetPointTest) {

    std::cout << "\tRequesting single setpoint with UUID..." << std::endl;

    SetPoint setpoint = BCAPI.getSetPoint(BASCLOUD_TEST_TENANT_UUID, BASCLOUD_TEST_STEPOINT_UUID);
    std::cout << "\t\tOK." << std::endl;

    std::cout << "\tSetpoint UUID: " << setpoint.getUUID() << std::endl;
    std::cout << "\tSetpoint Value: " << setpoint.getValue() << std::endl;
    std::cout << "\tSetpoint timestamp: " << fmt::format("{:%F %T}", fmt::localtime(setpoint.getTimestamp())) << std::endl;
    std::cout << "\tSetpoint created at: " << fmt::format("{:%F %T}", fmt::localtime(setpoint.getCreatedDate())) << std::endl;
    std::cout << "\tSetpoint updated at: " << fmt::format("{:%F %T}", fmt::localtime(setpoint.getLastUpdatedDate())) << std::endl;

    EXPECT_EQ(setpoint.getUUID(), BASCLOUD_TEST_STEPOINT_UUID);
    EXPECT_EQ(setpoint.getValue(), 2345.67);
    EXPECT_TRUE(setpoint.getTimestamp() > 0);
    EXPECT_TRUE(setpoint.getCreatedDate() > 0);
    EXPECT_TRUE(setpoint.getLastUpdatedDate() > 0);
}

TEST(BasicTests, CreateAndDeleteSetPointTest) {

    std::time_t currentDateTime = std::chrono::duration_cast<std::chrono::seconds>(std::chrono::system_clock::now().time_since_epoch()).count();

    SetPoint new_set = BCAPI.createSetPoint(BASCLOUD_TEST_TENANT_UUID, BASCLOUD_TEST_DEVICE_UUID, 2345.67, currentDateTime);
    std::cout << "\t\tOK." << std::endl;

    std::string new_uuid = new_set.getUUID();
    std::cout << "\t\tUUID: " << new_uuid << std::endl;

    EXPECT_EQ(new_set.getValue(), 2345.67);
    EXPECT_EQ(new_set.getTimestamp(), currentDateTime);
    // EXPECT_TRUE(new_set.getCreatedDate()-(2*60*60) >= currentDateTime); // TODO ML: this is wrong at the moment, for some reason the backend returns timestamps -2hours
    // EXPECT_TRUE(new_set.getLastUpdatedDate()-(2*60*60) >= currentDateTime);

    std::cout << "\tAgain Requesting all setpoints..." << std::endl; // Thus deactivating/inactivating them

    std::cout << "\tFrom: " << fmt::format("{:%F %T}", fmt::localtime(currentDateTime-(60*60*24))) << std::endl;
    std::cout << "\tUntil: " << fmt::format("{:%F %T}", fmt::localtime(currentDateTime+(60*60*24))) << std::endl;

    EntityCollection<SetPoint> setpoints = BCAPI.getSetPointsCollection(BASCLOUD_TEST_TENANT_UUID, {}, currentDateTime-(60*60*24), currentDateTime+(60*60*24), -1, -1, {}, -1, -1, [](const std::string& e, const std::string& j) {
            throw e;
        });
    std::cout << "\t\tOK." << std::endl;

    std::cout << "\tFound setpoints: " << setpoints.first.size() << std::endl;

    EXPECT_TRUE(setpoints.first.size() >= 1); // may no set points available

}

int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);

    authenticate();

    return RUN_ALL_TESTS();
}